impl/maven-di/src/main/java/org/apache/maven/di/impl/InjectorImpl.java

```diff
@@ -426,29 +425,14 @@ private void removeFromResolutionStack(Key<?> key) {
     }
 
     private static class SingletonScope implements Scope {
-        Map<Key<?>, java.util.function.Supplier<?>> cache = new ConcurrentHashMap<>();
+        final Map<Key<?>, java.util.function.Supplier<?>> cache = new ConcurrentHashMap<>();
 
         @Nonnull
         @SuppressWarnings("unchecked")
         @Override
         public <T> java.util.function.Supplier<T> scope(
                 @Nonnull Key<T> key, @Nonnull java.util.function.Supplier<T> unscoped) {
-            return (java.util.function.Supplier<T>)
-                    cache.computeIfAbsent(key, k -> new java.util.function.Supplier<T>() {
-                        volatile T instance;
-
-                        @Override
-                        public T get() {
-                            if (instance == null) {
-                                synchronized (this) {
-                                    if (instance == null) {
-                                        instance = unscoped.get();
-                                    }
-                                }
-                            }
-                            return instance;
-                        }
-                    });
+            return (java.util.function.Supplier<T>) cache.computeIfAbsent(key, k -> unscoped);
```

Comment:
Wow, don't change the logic here !  You're just completely breaking the _singleton scope_ which would not memoize the values anymore.   Your proposal just returns the original `unscoped` supplier, so `unscoped.get()` will be called every time the returned supplier is invoked, whereas the first one memoize the value so that it's only called once.

Reply:
ide suggestion.